### DISCUSSION TITLE
[golang]  project layout and package name linting

### DISCUSSION THREAD

**Discussion by Big-Iron-Cheems** [bufbuild/buf#3369] [Q&A] [answered]
First time using this tool to generate my sources from `.proto` files.
I am having some difficulties in understanding the layout that I'm supposed to use.
Below is my current state of the project, with an example file from it.

```yaml
# buf.yaml
version: v2

lint:
  use:
    - STANDARD
breaking:
  use:
    - FILE

modules:
  - path: cmd/test_lab/proto
```
```yaml
# buf.gen.yaml
version: v2

plugins:
  - local: protoc-gen-go
    out: cmd/test_lab/pb
    opt: paths=source_relative

inputs:
  - directory: cmd/test_lab/proto
```
```proto
// keyboard_message.proto
syntax = "proto3";

package cmd.test_lab.proto;  //< This is the part that the linter complains about

option go_package = "cmd/test_lab/pb";

message Keyboard {
  enum Layout {
    UNKNOWN = 0;
    QWERTY = 1;
    QWERTZ = 2;
    AZERTY = 3;
  }

  Layout layout = 1;
  bool backlit = 2;
}

```
```
# Project structure
.
├── buf.gen.yaml
├── buf.yaml
├── cmd
│   └── test_lab
│       ├── main.go
│       ├── pb
│       │   ├── keyboard_message.pb.go
│       │   ├── laptop_message.pb.go
│       │   ├── memory_message.pb.go
│       │   ├── processor_message.pb.go
│       │   ├── screen_message.pb.go
│       │   └── storage_message.pb.go
│       └── proto
│           ├── keyboard_message.proto
│           ├── laptop_message.proto
│           ├── memory_message.proto
│           ├── processor_message.proto
│           ├── screen_message.proto
│           └── storage_message.proto
├── go.mod
└── go.sum
```

While this builds fine when calling `buf generate`, I still get the following warning when I do `buf lint cmd/test_lab/proto/keyboard_message.proto`:
> Files with package "cmd.test_lab.proto" must be within a directory "cmd/test_lab/proto" relative to root but were in directory ".".

It is clear that this has to do with the package name and the position of the `.proto` files.
Let me know if I am misunderstanding how this is supposed to work, I am a complete newbie in this field.

**Comment by emcfarlane** [comment]
Hi @Big-Iron-Cheems, this is an issue with the path to the module as specified in the `buf.yaml` as `- path: cmd/test_lab/proto`. This path makes each file under `test_lab` at the '.' (root) directory. So another module importing this module would import the proto file as:
```
import "keyboard_message.proto"
```
This is probably not what you'd want.  You can remove the `modules:` section in the yaml to get the default path `.` which would make the file imported in other modules as:
```
import "cmd/test_lab/proto/keyboard_message.proto"
```
Have a look at the [style guide](https://buf.build/docs/best-practices/style-guide) for best practises and let me know if you have any issues!

**Reply by Big-Iron-Cheems** [reply]
Removing the `modules` section and adjusting the imports would result in this structure, where `cmd/test_lab/pb/cmd/test_lab/proto/keyboard_message.pb.go` doesn't look quite right.  
```
.
├── buf.gen.yaml
├── buf.yaml
├── cmd
│   └── test_lab
│       ├── main.go
│       ├── pb
│       │   └── cmd
│       │       └── test_lab
│       │           └── proto
│       │               ├── keyboard_message.pb.go
│       │               ├── laptop_message.pb.go
│       │               ├── memory_message.pb.go
│       │               ├── processor_message.pb.go
│       │               ├── screen_message.pb.go
│       │               └── storage_message.pb.go
│       └── proto
│           ├── keyboard_message.proto
│           ├── laptop_message.proto
│           ├── memory_message.proto
│           ├── processor_message.proto
│           ├── screen_message.proto
│           └── storage_message.proto
├── go.mod
└── go.sum
```

It would seem that most docs expect `proto` to be at the project root, and for it to mirror the folder structure in `cmd`.

**Reply by Big-Iron-Cheems** [reply, marked as answer]
![image](https://github.com/user-attachments/assets/64c5c3c4-dd07-477b-a8e5-be4fa2d75063)
Update (with image for clarity): this seems to work like I intended now:
- have everything in one `proto` folder
- have the gen code go in a `pb` folder
- no lint related to package name

I'm sure that from this example I'll be able to expand easily for any given structure.
Thanks for your input :+1:

**Reply by emcfarlane** [reply]
For generating the `go` code have a look at the source_relative option. You might find this [tutorial](https://buf.build/docs/generate/tutorial) helpful.
```
  - local: protoc-gen-go
    out: gen
    opt:
      - paths=source_relative
```

Edit: great I missed that you had applied it already. Glad you got it working!
